cmd/influxd/handler.go

```diff
@@ -1,42 +1,117 @@
 package main
 
 import (
+	"log"
 	"net/http"
+	"net/url"
 	"strings"
+
+	"github.com/influxdb/influxdb"
+	"github.com/influxdb/influxdb/httpd"
+	"github.com/influxdb/influxdb/messaging"
+	"github.com/influxdb/influxdb/raft"
 )
 
 // Handler represents an HTTP handler for InfluxDB node.
 // Depending on its role, it will serve many different endpoints.
 type Handler struct {
-	brokerHandler http.Handler
-	serverHandler http.Handler
+	Log    *raft.Log
+	Broker *influxdb.Broker
+	Server *influxdb.Server
+	Config *Config
 }
 
 // NewHandler returns a new instance of Handler.
-func NewHandler(bh, sh http.Handler) *Handler {
-	return &Handler{
-		brokerHandler: bh,
-		serverHandler: sh,
-	}
+func NewHandler() *Handler {
+	return &Handler{}
 }
 
 // ServeHTTP responds to HTTP request to the handler.
 func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	// Route raft and messaging paths to the broker.
-	if strings.HasPrefix(r.URL.Path, "/raft") || strings.HasPrefix(r.URL.Path, "/messaging") {
-		if h.brokerHandler == nil {
-			http.NotFound(w, r)
-			return
+	if strings.HasPrefix(r.URL.Path, "/raft") {
+		h.serveRaft(w, r)
+		return
+	}
+	if strings.HasPrefix(r.URL.Path, "/messaging") {
+		h.serveMessaging(w, r)
+		return
+	}
+
+	h.serveData(w, r)
+}
+
+// serveMessaging responds to broker requests
+func (h *Handler) serveMessaging(w http.ResponseWriter, r *http.Request) {
+	// If we're running a broker, handle the broker endpoints
+	if h.Broker != nil {
+		mh := &messaging.Handler{
+			Broker:      h.Broker.Broker,
+			RaftHandler: &raft.Handler{Log: h.Log},
 		}
+		mh.ServeHTTP(w, r)
+		return
+	}
+
+	if h.Server == nil {
+		log.Println("no broker or server configured to handle messaging endpoints")
+		http.Error(w, "server unavailable", http.StatusServiceUnavailable)
+		return
+	}
+
+	// Redirect to a valid broker to handle the request
+	h.redirect(h.Server.BrokerURLs(), w, r)
+}
 
-		h.brokerHandler.ServeHTTP(w, r)
+// serveRaft responds to raft requests.
+func (h *Handler) serveRaft(w http.ResponseWriter, r *http.Request) {
+	if h.Log != nil {
+		rh := raft.Handler{Log: h.Log}
+		rh.ServeHTTP(w, r)
 		return
 	}
 
-	// Route all other paths to the server.
-	if h.serverHandler == nil {
-		http.NotFound(w, r)
+	if h.Server == nil {
```

Comment:
In general, I feel like this error checking should take place first thing in the function. Check that there is going to be at least one non-nil resource that will handle the call, then proceed to handle it.